### PR TITLE
implement stateless resets

### PR DESCRIPTION
**Pull request by marten-seemann**
This PR changes the server state machine such that sending HelloRetryRequests can be performed statelessly.

In non-blocking mode, `Handshake()` will return a new alert, `AlertStatelessRetry`. When receiving this return, an app can discard all mint-related state (i.e. delete `Conn` garbage collected). When not using non-blocking mode, the handshake is performed as stateless as possible, at no point the initial ClientHello and the HelloRetryRequest will be stored.

The cookie that is sent to the client in the HelloRetryRequest is composed of two parts: In the mint-part of the cookie, we store the hash of the initial ClientHello. The second part can be used to store an application-defined cookie (used when `Config.CookieHandler` is set). The application has to provide a way to encrypt this cookie (it should use authenticated encryption for this). This can be done by setting `Config.CookieSource`. In blocking mode, we default to use AES-GCM if no `CookieSource` is set.

@ekr, @bifurcation: Please tell me what you think about this. The stateless handshake works for me with this PR, but I can't guarantee I didn't miss anything subtle here. My current understanding of the TLS 1.3 draft doesn't go a lot farther than what I need for my mint changes and the QUIC integration.

### CODE REVIEW

**Comment by ekr** [comment]
@marten-seemann this seems to fail in CI...

File: alert.go [hunk 1]
```diff
@@ -82,6 +83,7 @@ var alertText = map[Alert]string{
 	AlertUnknownPSKIdentity:          "unknown PSK identity",
 	AlertNoApplicationProtocol:       "no application protocol",
 	AlertNoRenegotiation:             "no renegotiation",
+	AlertStatelessRetry:              "stateless retry",
```
Comment (ekr) [inline alert.go]: This isn't your fault, but I wonder if we should give these non-alerts a special name so we don't acccidentally try to send them.

File: conn.go [hunk 2]
```diff
@@ -74,9 +74,16 @@ type Config struct {
 	AllowEarlyData     bool
 	// Require the client to echo a cookie.
 	RequireCookie bool
+	// TODO: fix comment here: no more default cookie handler
```
Comment (ekr) [inline conn.go]: This TODO seems like it needs fixing.

File: conn.go [hunk 3]
```diff
@@ -74,9 +74,16 @@ type Config struct {
...
 	// If cookies are required and no CookieHandler is set, a default cookie handler is used.
 	// The default cookie handler uses 32 random bytes as a cookie.
-	CookieHandler     CookieHandler
+	CookieHandler CookieHandler
+	// The CookieSource is used to encrypt / decrypt cookies.
+	// TODO: implement a check for that
+	// If non-blocking mode is used, and cookies are required, this field has to be set.
+	// TODO: implement that
```
Comment (ekr) [inline conn.go]: These TODOs also seem to need fixing

File: conn.go [hunk 4]
```diff
@@ -642,6 +648,19 @@ func (c *Conn) HandshakeSetup() Alert {
...
+		if c.config.RequireCookie && c.config.CookieSource == nil {
+			logf(logTypeHandshake, "RequireCookie set, but no CookieSource provided. Using default cookie source. Stateless Retry not possible.")
+			if c.config.NonBlocking {
+				logf(logTypeHandshake, "Not possible in non-blocking mode.")
+				return AlertInternalError
+			}
+			var err error
+			caps.CookieSource, err = newDefaultCookieSource()
+			if err != nil {
+				logf(logTypeHandshake, "Error initializing client state: %v", alert)
```
Comment (ekr) [inline conn.go]: Should this say server state?

File: conn.go [hunk 5]
```diff
@@ -687,6 +706,7 @@ func (c *Conn) Handshake() Alert {
 
 	for !connected {
 		// Read a handshake message
+		logf(logTypeHandshake, "readmessage")
```
Comment (ekr) [inline conn.go]: Maybe make this |logTypeVerbose| and make the text clearer.

File: conn.go [hunk 6]
```diff
@@ -720,6 +738,13 @@ func (c *Conn) Handshake() Alert {
 		c.hState = state
 		logf(logTypeHandshake, "state is now %s", c.GetHsState())
 
+		// if alert == AlertStatelessRetry {
+		// 	logf(logTypeHandshake, "Doing a retry")
+		// 	if c.config.NonBlocking {
+		// 		return AlertStatelessRetry
+		// 	}
+		// }
```
Comment (ekr) [inline conn.go]: You should remove this code.

File: cookie_source.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func deriveKey(secret []byte) ([]byte, error) {
+	r := hkdf.New(sha256.New, secret, nil, []byte("mint TLS 1.3 cookie token key"))
+	key := make([]byte, tokenKeySize)
+	if _, err := io.ReadFull(r, key); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
```
Comment (ekr) [inline cookie_source.go]: I'm not sure it's necessary to have an HKDF stage here. You are feeding in random bits, right?

File: cookie_source.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	return &defaultCookieSource{aead: aead}, nil
+}
+
+func (s *defaultCookieSource) NewToken(data []byte) ([]byte, error) {
+	nonce := make([]byte, tokenNonceSize)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	return s.aead.Seal(nonce, nonce, data, nil), nil
```
Comment (ekr) [inline cookie_source.go]: I don't think this is the crypto structure you want here. The issue is that you have a non-trivial chance of nonce-reuse at 2^96 because of birthday, which is catastrophic with GCM. You should either use an integer nonce (that won't repeat within any reasonable bound) or instead generate a longer nonce (256 bits is good) and use it to derive both the per-message key and per-message nonce

File: server-state-machine.go [hunk 9]
```diff
@@ -57,13 +60,18 @@ import (
...
 //  CONNECTED					StoreTicket || (RekeyIn; [RekeyOut])
 
+// A cookie can be sent to the client in a HRR.
+// It contains two fields:
+// 1. The MintCookie: This field is used by mint itself to store the hash of initial client hello.
+// 2. The ApplicationCookie: This opaque value can be provided by the application (by setting a Config.CookieHandler)
+type cookie struct {
+	MintCookie        []byte `tls:"head=2"`
+	ApplicationCookie []byte `tls:"head=2"`
+}
```
Comment (ekr) [inline server-state-machine.go]: In NSS, we also start the negotiated parameters and double-check that they don't change

File: server-state-machine.go [hunk 10]
```diff
@@ -129,9 +139,29 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
...
-	if state.Caps.RequireCookie && state.cookieSent && !state.Caps.CookieHandler.Validate(state.conn, clientCookie.Cookie) {
-		logf(logTypeHandshake, "[ServerStateStart] Cookie mismatch")
-		return nil, nil, AlertAccessDenied
+	// The client sent a cookie. So this is probably the second ClientHello (sent as a response to a HRR)
+	var firstClientHello *HandshakeMessage
+	if clientSentCookie {
+		plainCookie, err := state.Caps.CookieSource.DecodeToken(clientCookie.Cookie)
+		if err != nil {
+			logf(logTypeHandshake, fmt.Sprintf("[ServerStateStart] Error decoding token [%v]", err))
+			return nil, nil, AlertAccessDenied
```
Comment (ekr) [inline server-state-machine.go]: I don't think you want AccessDenied here, but rather DecryptError or IllegalParameter

File: server-state-machine.go [hunk 11]
```diff
@@ -129,9 +139,29 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
...
+		}
+		cookie := &cookie{}
+		if _, err := syntax.Unmarshal(plainCookie, cookie); err != nil {
+			logf(logTypeHandshake, fmt.Sprintf("[ServerStateStart] Error unmarshaling cookie [%v]", err))
+			return nil, nil, AlertAccessDenied
+		}
```
Comment (ekr) [inline server-state-machine.go]: This should be internal_error, because it can't happen.

File: server-state-machine.go (hunk 10 above)
Comment (ekr) [reply server-state-machine.go]: I would put a TODO here because if you were to change your key, this wouldn't work.

File: server-state-machine.go [hunk 12]
```diff
@@ -142,26 +172,51 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
...
-		}
 
-		chTrunc, err := ch.Truncated()
+	if len(clientPSK.Identities) > 0 {
+		canDoPSK, selectedPSK, psk, params, err = PSKNegotiation(clientPSK.Identities, state.Caps.PSKs)
 		if err != nil {
-			logf(logTypeHandshake, "[ServerStateStart] Error computing truncated ClientHello [%v]", err)
-			return nil, nil, AlertDecodeError
+			logf(logTypeHandshake, "[ServerStateStart] Error in PSK negotiation [%v]", err)
+			return nil, nil, AlertInternalError
```
Comment (ekr) [inline server-state-machine.go]: I am surprised this is internal error. @bifurcation ?

File: server-state-machine.go [hunk 13]
```diff
@@ -142,26 +172,51 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
...
+			if err != nil {
+				logf(logTypeHandshake, "[ServerStateStart] Error computing truncated ClientHello [%v]", err)
+				return nil, nil, AlertDecodeError
+			}
+			ctxHash.Write(chTrunc)
+
+			binder := computeFinishedData(params, binderKey, ctxHash.Sum(nil))
+			if !bytes.Equal(binder, clientPSK.Binders[selectedPSK].Binder) {
+				logf(logTypeNegotiation, "Binder check failed for identity %x; [%x] != [%x]", psk.Identity, binder, clientPSK.Binders[selectedPSK].Binder)
+				return nil, nil, AlertInternalError
```
Comment (ekr) [inline server-state-machine.go]: This seems like decrypt_error.

File: server-state-machine.go [hunk 14]
```diff
@@ -175,61 +230,66 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
...
+			if state.Caps.CookieHandler == nil { // if Config.RequireCookie is set, but no CookieHandler was provided, we definitely need to send a cookie
+				shouldSendHRR = true
+			} else { // if the CookieHandler was set, we just send a cookie when the application provides one
+				var err error
+				appCookie, err = state.Caps.CookieHandler.Generate(state.conn)
+				if err != nil {
+					logf(logTypeHandshake, "[ServerStateStart] Error generating cookie [%v]", err)
+					return nil, nil, AlertInternalError
+				}
+				shouldSendHRR = appCookie != nil
```
Comment (ekr) [inline server-state-machine.go]: Is this idiomatic go? I would have done if....

**Comment by marten-seemann** [comment]
Hi, @ekr thanks for the review! Seems like I forgot a bunch of TODOs and a few debug statements. I removed those, and added a bunch of documentation about `Cookie`s, `CookieSource`s and `CookieHandler`.
The reason the tests were failing is that CircleCI apparently uses Go 1.8 if no config file is provided, and the test helper functions (`t.Helper()`) require Go 1.9. I added a basic config file.

I really like the idea of storing the negotiated parameters in the cookie as well. By doing so, I was able to remove a rather ugly hack that was necessary to make PSK mode work. Is there anything else besides the `CipherSuite` that we would want to store in the cookie?

The code for the cookie source is what we copied from Google's QUIC implementation 1.5 years ago or so. I changed the cookie source to use a nonce size of 32 byte, as you suggested, and also removed the HKDF. I'm not sure though why're suggesting using a per-message key, I thought using the same key will be ok as long as the nonce is not reused. Can you please explain?
I also made the `DefaultCookieSource` public. When using non-blocking mode, an app will have to provide its own implementation for the `CookieSource`. The `DefaultCookieSource` should work fine in easy use cases, such as running a single server (in a multi-server deployment you'll need some shared cookie source anyway, and probably also implement some key rotation mechanism).

**Comment by ekr** [comment]
The issue with the nonce here is that AES-GCM fails catastrophically if you have nonce reuse. If you use a random nonce, you get a non-trivial probability of reuse with enough connections (the 50% mark is at 2^{48} but the risk is unacceptably high well below that). Unfortunately, just making the nonce bigger doesn't help because AES-GCM hashes a >96 bit nonce down to 96 bits.

If you want to use GCM, you can guarantee a unique key/nonce pair by generating a longer nonce and then using HKDF to turn the master key into a key/nonce. Because the output is longer, the chance of collision is reduced.

**Comment by marten-seemann** [comment]
I just pushed a fixed. We're now creating a new key (which is HKDF derived using the nonce as a salt) for every nonce.

File: cookie_source.go [hunk 15]
```diff
@@ -11,63 +11,68 @@ import (
...
+const cookieNonceSize = 32
+
+// The DefaultCookieSource is a simple implementation for the CookieSource.
+type DefaultCookieSource struct {
+	secret []byte
 }
 
-const tokenKeySize = 16
-const tokenNonceSize = 16
+var _ CookieSource = &DefaultCookieSource{}
```
Comment (ekr) [inline cookie_source.go]: I am probably not enough of a Golang wizard, but why is this neded.

File: cookie_source.go [hunk 16]
```diff
@@ -11,63 +11,68 @@ import (
...
+	nonce := make([]byte, cookieNonceSize)
 	if _, err := rand.Read(nonce); err != nil {
 		return nil, err
 	}
-	return s.aead.Seal(nonce, nonce, data, nil), nil
+	aead, err := s.createAEAD(nonce)
+	if err != nil {
+		return nil, err
+	}
+	return aead.Seal(nonce, nonce, data, nil), nil
```
Comment (ekr) [inline cookie_source.go]: I don't think this is quite what you want. Rather, you want to do:

```
r := make([]byte, 32)
rand.Read(r)
key = hkdf(secret, R, []byte("mint cookie key")) // key length
nonce = hkdf(secret, R, []byte("mint cookie nonce")) // 12 bytes
```

And then you can use nonce the way you have it here.

Comment (ekr) [reply cookie_source.go]: Or, if you prefer you can just read the key and the nonce as successive values from the hkdf.

File: cookie_source.go [hunk 17]
```diff
@@ -11,63 +11,68 @@ import (
...
+	key := make([]byte, 32)
 	if _, err := io.ReadFull(r, key); err != nil {
 		return nil, err
 	}
-	return key, nil
+	c, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	return cipher.NewGCMWithNonceSize(c, cookieNonceSize)
```
Comment (ekr) [inline cookie_source.go]: You want NewGCM()

File: cookie_source_test.go [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+		decoded, err := cs.DecodeToken(token)
+		assertNotError(t, err, "decoding the token failed")
+		assertDeepEquals(t, cookie, decoded)
+	})
+
+	t.Run("handling invalid tokens", func(t *testing.T) {
+		_, err := cs.DecodeToken([]byte("too short"))
+		assertError(t, err, "it should reject too short tokens")
+		_, err = cs.DecodeToken(append(bytes.Repeat([]byte{0}, cookieNonceSize), []byte("invalid token")...))
+		assertError(t, err, "it should reject invalid tokens")
```
Comment (ekr) [inline cookie_source_test.go]: It would also be good to have one with a bogus nonce, e.g., by stomping the first byte.

File: cookie_source.go (hunk 15 above)
Comment (marten-seemann) [reply cookie_source.go]: Strictly speaking this is not needed. It makes the compiler check that `DefaultCookieSource` actually implements the `CookieSource` interface. Since it's an assignment to `_`, it will be removed during compilation.

File: cookie_source.go (hunk 16 above)
Comment (marten-seemann) [reply cookie_source.go]: How is that supposed to work? When decoding the cookie, I don't know `r` any more, and I can't create the AEAD to open the cookie.

Comment (ekr) [reply cookie_source.go]: The cookie is r || aead.Seal(blah...)

Right now you are just using nonce == r, so the difference I am proposing is you hkdf r to make the nonce.

**Comment by marten-seemann** [comment]
@ekr, I implemented the changes for the `CookieSource` as you suggested. Please have another look.

**Comment by ekr** [comment]
@marten-seemann I think you forgot to push

**Comment by marten-seemann** [comment]
I don't think so. I rebased and amended the commit.

**Comment by ekr** [comment]
Oh, ok. Will take a look

File: cookie_source.go [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+func (s *DefaultCookieSource) createAEAD(nonce []byte) (cipher.AEAD, []byte, error) {
+	h := hkdf.New(sha256.New, s.secret, nonce, []byte("mint cookie source"))
+	key := make([]byte, 32) // use a 32 byte key, in order to select AES-256
+	if _, err := io.ReadFull(h, key); err != nil {
+		return nil, nil, err
+	}
+	aeadNonce := make([]byte, 12)
+	if _, err := io.ReadFull(h, aeadNonce); err != nil {
+		return nil, nil, err
+	}
```
Comment (ekr) [inline cookie_source.go]: This looks fine to me. Assuming you didn't change anything else, I think this is good to Go. @bifurcation ??

Comment (marten-seemann) [reply cookie_source.go]: I didn't change anything else. This should be ready to merge now.

File: .circleci/config.yml [hunk 20]
```diff
@@ -0,0 +1,10 @@
+version: 2
```
Comment (bifurcation) [inline .circleci/config.yml]: Unless this change is necessary for this functionality, please put it in a separate PR.

Comment (marten-seemann) [reply .circleci/config.yml]: This was the first commit of this PR (03b32d6dfee03aaed5b3759a24fe5ef8d54cd943). I openend #147, containing just this one commit.

**Comment by marten-seemann** [comment]
@bifurcation: Is there anything else you need from me to merge this PR?

File: cookie_source.go [hunk 21]
```diff
@@ -0,0 +1,86 @@
+package mint
```
Comment (bifurcation) [inline cookie_source.go]: For consistency with the other files, this should be `cookie-source.go` (reserving underscores for `_test`).

File: conn.go [hunk 22]
```diff
@@ -74,9 +78,16 @@ type Config struct {
 	AllowEarlyData     bool
 	// Require the client to echo a cookie.
 	RequireCookie bool
-	// If cookies are required and no CookieHandler is set, a default cookie handler is used.
-	// The default cookie handler uses 32 random bytes as a cookie.
-	CookieHandler     CookieHandler
+	// A CookieHandler can be used to set and validate a cookie.
+	// The cookie returned by the CookieHandler will be part of the cookie sent on the wire, and endoced using the CookieSource.
```
Comment (bifurcation) [inline conn.go]: Nit: "encoded"

File: conn.go [hunk 23]
```diff
@@ -74,9 +78,16 @@ type Config struct {
...
+	// If no CookieHandler is set, mint will always send a cookie.
+	// The CookieHandler can be used to decide on a per-connection basis, if a cookie should be sent.
+	CookieHandler CookieHandler
+	// The CookieSource is used to encrypt / decrypt cookies.
+	// It should make sure that the Cookie cannot be read and tampered with by the client.
+	// If non-blocking mode is used, and cookies are required, this field has to be set.
+	// In blocking mode, a default cookie source is used, if this is unused.
+	CookieSource      CookieSource
```
Comment (bifurcation) [inline conn.go]: Nit: `CookieSource` doesn't seem like the right name for something whose job is to protect / unprotect cookie values.  `CookieProtection` or something like that?

**Comment by marten-seemann** [comment]
@bifurcation: I rebased this PR on top of the current master, fixed the type, and renamed the `CookieSource` to `CookieProtector`.

**Comment by bifurcation** [comment]
@marten-seemann Thanks! Merging and moving on to #145 ...
